File: 14-dockingData/go/challenge/partOne.go
package challenge

import (
	"strconv"
	"strings"
)

func applyMask(number int, mask string) int {
	mask = strings.ToLower(mask)
	value := numberToBase(number, 2)
	value = zfill(value, len(mask)-len(value))
	var combi string
	for i := 0; i < len(value); i += 1 {
		v := string(value[i])
		m := string(mask[i])

		if m == "x" {
			combi += v
		} else {
			combi += m
		}
	}
	ix, err := strconv.ParseInt(combi, 2, 64)
	if err != nil {
		panic(err)
	}
	return int(ix)
}

func PartOne(instr string) int {
	inputInstructions := parse(instr)
	memory := make(map[int]int)

	for _, instruction := range inputInstructions {
		memory[instruction.Address] = applyMask(instruction.Value, instruction.Mask)
	}

	var sigma int
	for _, v := range memory {
		sigma += v
	}
	return sigma
}

File: challenges/2020/14-dockingData/python/partOne.py
from common import *


def apply_mask(number: int, mask: str) -> int:
    # There has to be a better way to do this... but oh well
    # it works, after all
    value = number_to_base(number, 2).zfill(len(mask))
    combi = ""
    for (v, m) in zip(value, mask.lower()):
        if m == "x":
            combi += v
        else:
            combi += m
    return int(combi, 2)


def partOne(instr: str) -> int:
    input_instructions = parse(instr)
    memory = {}

    for instruction in input_instructions:
        memory[instruction.address] = apply_mask(instruction.value, instruction.mask)

    sigma = 0
    for key in memory:
        sigma += memory[key]

    return sigma
